en/_posts/2016-07-04-local-stateful.md

```diff
@@ -0,0 +1,115 @@
+---
+layout: post
+title:  "Local stateful"
+categories: docker
+author: tgermain
+lang: en
+---
+
+
+
+This tutorial explains how local stateful and volumes work.
+
+
+
+# Volumes :
+
+The current version of volumes works exactly as docker host volume : the idea is to share a directory or a file between the host and the container.
+
+Example :
+
+```
+host $  mkdir -p ~/data/toto
+host $  touch ~/data/toto/hello
+host $  docker run -ti -v ~/data/toto:/data/ alpine sh
+container $ ls /data
+hello
+container $ touch /data/fromTheOtherSide
+host $ ls ~/data/toto/
+fromTheOtherSide  hello
+```
+
+Here we create a directory `~/data/toto` on the host, let's call it the **hostPath**. We create a file *hello* in it.
+
+When we run the container, we tell docker to map `~/data/toto` aka the **hostPath** to `/data/`, the **containerPath**, inside the container.
+
+The binding is bidirectionnal : if the container creates a file `/data/fromTheOtherSide`, we are able to see it from the host.
+
+
+
+To protect our infrastructure, we restrain what a user can mount into its containers.
+To do that we [chroot](https://en.wikipedia.org/wiki/Chroot) any **containerPath**.
+
+
+Example:
+
+For a user toto, this volume :
+
+```
+"volumes": [
+    {
+          "containerPath": "/log",
+          "hostPath": "/var/log",
+    }
+]
+```
+
+is tranformed into :
+
+```
+"volumes": [
+    {
+          "containerPath": "/log",
+          "hostPath": "/containerdata/user-toto/var/log",
+    }
+]
+```
+
+
+# Problems
+
+## Chrooting issue
+
+This can cause troubles, especially if you try to mount a file from the host and read it from the container :
+
+We want to mount the file `/etc/localtime` from the host to `/etc/localtime` in the container to get the same time settings as the host.
+
+The **hostPath** is modified from `/etc/localtime` to `/containerdata/user-toto/etc/localtime`.
+
+
+This path does not exist on the host, when the container starts docker creates it as a directory, it's the default behaviour.
+If the code inside the container tries to read  `/etc/localtime` (the **containerPath**) as a file, it fails because it's a directory on the host.
+
+
+
+If you run into that kind of problem, you can try to troubleshoot it in the "Debug" tab of your faulty application:
+
+```
+Message executor
+
+    failed to run container: docker failed to start container  0cf43b13cbd46912d43df6f760a21d5fb28e759ecd74ed240d156fef96d34347: Error  response from daemon: rpc error: code = 2 desc = "oci runtime error:  could not synchronise with container process: not a directory"
+```
+
+The message is cryptic but show that a directory was treated as a file and it didn't go well.
```

Comment:
"This message basically means..." "cryptic" tends to suggest mystery hence to discourage from learning.